go/apps/ctrl/services/acme/certificate_workflow.go

```diff
@@ -0,0 +1,129 @@
+package acme
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/go-acme/lego/v4/certificate"
+	"github.com/go-acme/lego/v4/lego"
+	vaultv1 "github.com/unkeyed/unkey/go/gen/proto/vault/v1"
+	"github.com/unkeyed/unkey/go/pkg/db"
+	"github.com/unkeyed/unkey/go/pkg/hydra"
+	"github.com/unkeyed/unkey/go/pkg/otel/logging"
+	pdb "github.com/unkeyed/unkey/go/pkg/partition/db"
+	"github.com/unkeyed/unkey/go/pkg/vault"
+)
+
+// CertificateChallenge tries to get a certificate from Let's Encrypt
+type CertificateChallenge struct {
+	db          db.Database
+	partitionDB db.Database
+	logger      logging.Logger
+	acmeClient  *lego.Client
+	vault       *vault.Service
+}
+
+type CertificateChallengeConfig struct {
+	DB          db.Database
+	PartitionDB db.Database
+	Logger      logging.Logger
+	AcmeClient  *lego.Client
+	Vault       *vault.Service
+}
+
+// NewCertificateChallenge creates a new certificate challenges workflow instance
+// and ensures that we have a valid ACME User
+func NewCertificateChallenge(config CertificateChallengeConfig) *CertificateChallenge {
```

Comment:
since you have been writing a workflow from scratch, how did it feel? 
what can we improve?

Reply:
I think we should add an indompoency id to a workflow run so if I spawn multiple tasks with the same one only the first one gets ran for a cooldown of XX duration